### PR TITLE
Refactor BibtexEntryType

### PR DESCRIPTION
**Pull request by stefan-kolb**
This is a massive refactoring of BibtexEntryTypes.

# Changes
- Divided EntryTypes into subclasses BibtexEbtryTypes, IEEETranEntryTpes, BiblatexEntryTypes and introduced a new EntryType interface.
- EntryTypes are now clustered per group on the 'add new entry' GUI screen.
- BibtexTypes and IEEETranTypes required field checking is consistent now.
- Entry type 'Other' is not selectable anymore as it is no real entry type.
- Removed getUtilityFields method from EntryType
- Removed getRequiredFieldsForCustomization as its result is equal with getRequiredFields now
- Removed isVisibleAtNewEntryDialog, visibility is controlled via an internal List for each EntryTypes collection

Further TODOs see #366

@JabRef/developers Everyone, please take a close look before merging this into master.

### CODE REVIEW

File: src/main/java/net/sf/jabref/model/entry/BibtexEntryType.java [hunk 1]
```diff
@@ -35,23 +35,23 @@
...
 
 /**
- * Provides a list of known entry types
- * <p>
- * The list of optional and required fields is derived from http://en.wikipedia.org/wiki/BibTeX#Entry_types
+ * Abstract base class for all entry types.
  */
 public abstract class BibtexEntryType implements Comparable<BibtexEntryType> {
+    private List<String> requiredFields;
+    private List<String> optionalFields;
```
Comment (simonharrer) [inline src/main/java/net/sf/jabref/model/entry/BibtexEntryType.java]: final?

File: src/main/java/net/sf/jabref/model/entry/BibtexEntryType.java [hunk 2]
```diff
@@ -35,23 +35,23 @@
 import net.sf.jabref.model.database.BibtexDatabase;
 
 /**
- * Provides a list of known entry types
- * <p>
- * The list of optional and required fields is derived from http://en.wikipedia.org/wiki/BibTeX#Entry_types
+ * Abstract base class for all entry types.
  */
 public abstract class BibtexEntryType implements Comparable<BibtexEntryType> {
```
Comment (simonharrer) [inline src/main/java/net/sf/jabref/model/entry/BibtexEntryType.java]: Idea: secondary optional only in biblatex, classes should reflect that. Boolean method if current class is biblatex class? Maybe interface + abstract classes for bibtex/biblatex?

File: src/main/java/net/sf/jabref/model/entry/BibtexEntryType.java [hunk 3]
```diff
@@ -82,23 +82,17 @@ void addAllRequired(String... fieldNames) {
         return optionalFields.stream().filter(field -> !isPrimary(field)).collect(Collectors.toList());
     }
 
-    public abstract boolean hasAllRequiredFields(BibtexEntry entry, BibtexDatabase database);
-
     public String[] getUtilityFields() {
```
Comment (simonharrer) [inline src/main/java/net/sf/jabref/model/entry/BibtexEntryType.java]: only used once in the DBExporter, inline and remove.

Comment (koppor) [inline src/main/java/net/sf/jabref/model/entry/BibtexEntryType.java]: I think, this method is useful also in other cases. I remember that there was a pattern for that: data objects should not carry "extended" methods, but a visitor or something similar should be used. `BibtexEntryTypeUtil`?

Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/model/entry/BibtexEntryType.java]: I don't understand that.

**Comment by stefan-kolb** [comment]
@simonharrer Maybe you could resolve the rendering of multiple stacked GridBadLayouts. I don't know how to do this.

**Comment by simonharrer** [comment]
Hm, GridBagLayouts are not my special area of expertise. Where are the rendering issues exactly?

File: src/test/java/net/sf/jabref/model/entry/BibtexEntryTests.java [hunk 4]
```diff
@@ -48,4 +51,49 @@ public void testGetPublicationDate() {
...
+        e.setField("title", "abc");
+        e.setField("journal", "abc");
+        List<String> requiredFields = new ArrayList<>();
+
+        // XOR required
+        requiredFields.add("journal/year");
+        Assert.assertTrue(e.allFieldsPresent(requiredFields, null));
+
+        requiredFields.add("year/address");
+        Assert.assertFalse(e.allFieldsPresent(requiredFields, null));
```
Comment (simonharrer) [inline src/test/java/net/sf/jabref/model/entry/BibtexEntryTests.java]: methods should start with a verb, e.g., areAllFieldsPresent

**Comment by koppor** [comment]
`@Other` is used in many bib files. Including my personal `bib` file. We have to offer a migration path. Currently, the entry editor does not open, because of `java.util.Objects.requireNonNull` (see below). One cannot quit JabRef after this error.

With [othertype.bib](https://github.com/JabRef/jabref/blob/94712c96b0069734b33c8ecff4b293c052bf2186/src/test/resources/testbib/othertype.bib), I get the error message `unknown entry type: other:test`. Did the parsing of the lines changed? Did the parsing change? Why is the colon (`:`) not recognized any more as separator?

In general: JabRef should still be able to handle arbitrary custom types. Even if not configured in the bib file. I heard from some persons that they manage contacts and we cannot forsee all customized types. To test that, I added [reallyunknowntype.bib](https://github.com/JabRef/jabref/blob/94712c96b0069734b33c8ecff4b293c052bf2186/src/test/resources/testbib/reallyunknowntype.bib).

```
Exception in thread "AWT-EventQueue-0" java.lang.ExceptionInInitializerError
	at net.sf.jabref.gui.entryeditor.EntryEditor.updateSource(EntryEditor.java:571)
	at net.sf.jabref.gui.entryeditor.EntryEditor.setupSourcePanel(EntryEditor.java:555)
	at net.sf.jabref.gui.entryeditor.EntryEditor.<init>(EntryEditor.java:194)
	at net.sf.jabref.gui.BasePanel.getEntryEditor(BasePanel.java:2075)
	at net.sf.jabref.gui.MainTableSelectionListener.editSignalled(MainTableSelectionListener.java:241)
	at net.sf.jabref.gui.MainTableSelectionListener.mouseClicked(MainTableSelectionListener.java:284)
	at java.awt.AWTEventMulticaster.mouseClicked(AWTEventMulticaster.java:270)
	at java.awt.Component.processMouseEvent(Component.java:6528)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
	at java.awt.Component.processEvent(Component.java:6290)
	at java.awt.Container.processEvent(Container.java:2234)
	at java.awt.Component.dispatchEventImpl(Component.java:4881)
	at java.awt.Container.dispatchEventImpl(Container.java:2292)
	at java.awt.Component.dispatchEvent(Component.java:4703)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4898)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4542)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4462)
	at java.awt.Container.dispatchEventImpl(Container.java:2278)
	at java.awt.Window.dispatchEventImpl(Window.java:2750)
	at java.awt.Component.dispatchEvent(Component.java:4703)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:75)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.awt.EventQueue$4.run(EventQueue.java:729)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:75)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: java.lang.NullPointerException
	at java.util.Objects.requireNonNull(Objects.java:203)
	at java.util.Arrays$ArrayList.<init>(Arrays.java:3813)
	at java.util.Arrays.asList(Arrays.java:3800)
	at net.sf.jabref.model.entry.CustomEntryType.getRequiredFields(CustomEntryType.java:91)
	at net.sf.jabref.model.entry.EntryType.getRequiredFieldsFlat(EntryType.java:44)
	at net.sf.jabref.bibtex.BibtexEntryWriter.<clinit>(BibtexEntryWriter.java:72)
	... 38 more
```

**Comment by koppor** [comment]
The closing seems to be prevented by

```
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
        at net.sf.jabref.model.entry.CustomEntryType.getRequiredFieldsString(CustomEntryType.java:112)
        at net.sf.jabref.JabRefPreferences.storeCustomEntryType(JabRefPreferences.java:1338)
        at net.sf.jabref.logic.CustomEntryTypesManager.saveCustomEntryTypes(CustomEntryTypesManager.java:44)
        at net.sf.jabref.gui.JabRefFrame.tearDownJabRef(JabRefFrame.java:812)
        at net.sf.jabref.gui.JabRefFrame.quit(JabRefFrame.java:917)
        at net.sf.jabref.gui.JabRefFrame$CloseAction.actionPerformed(JabRefFrame.java:1759)
        at net.sf.jabref.gui.JabRefFrame$3.windowClosing(JabRefFrame.java:563)
```

**Comment by lenhard** [comment]
Is the closing bug fixed with e725a5ee2f4a9e33ac616d38c2104f75aa545655 ?

**Comment by stefan-kolb** [comment]
I'll take a look after the lecture. Thanks for the Testing!
Am 19.11.2015 5:45 nachm. schrieb "Jörg Lenhard" <notifications@github.com>:

**Comment by koppor** [comment]
Saving works now.

**Comment by stefan-kolb** [comment]
I don't get any exceptions. Except that you get the message that the type is unknown which is expected. What should change in your opinion? I also got these messages before the current change.

**Comment by koppor** [comment]
Fixed in `master` :tiger2: :+1:

**Pull request merged**